### ISSUE TITLE
OpenSC does not recognize Latvian eID card

### ISSUE THREAD

**Issue by mouse07410** [OpenSC/OpenSC#1275]
### Problem Description

```
$ opensc-tool -aln
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             SCM Microsystems Inc. SCR 355
Using reader with a card: SCM Microsystems Inc. SCR 355
3b:dd:18:00:81:31:fe:45:90:4c:41:54:56:49:41:2d:65:49:44:90:00:8c
Unsupported card
$
```

Debug log: 
[latID.debug.log.txt](https://github.com/OpenSC/OpenSC/files/1756742/latID.debug.log.txt) 
obtained via `OPENSC_DEBUG=9 PKCS11SPY_OUTPUT=~/latID.spy.txt pkcs11-tool --module /Library/OpenSC/lib/pkcs11-spy.dylib -LIT`

SPY doesn't tell anything useful:
```
*************** OpenSC PKCS#11 spy *****************
Loaded: "/Library/OpenSC/lib/opensc-pkcs11.dylib"

0: C_GetFunctionList
2018-02-25 16:29:48.504
Returned:  0 CKR_OK

1: C_Initialize
2018-02-25 16:29:48.505
[in] pInitArgs = 0x0
Returned:  0 CKR_OK

2: C_GetInfo
2018-02-25 16:29:49.243
[out] pInfo: 
      cryptokiVersion:         2.20
      manufacturerID:         'OpenSC Project                  '
      flags:                   0
      libraryDescription:     'OpenSC smartcard framework      '
      libraryVersion:          0.17
Returned:  0 CKR_OK

3: C_GetSlotList
2018-02-25 16:29:49.244
[in] tokenPresent = 0x1
[out] pSlotList: 
Count is 0
[out] *pulCount = 0x0
Returned:  0 CKR_OK

4: C_GetSlotList
2018-02-25 16:29:49.741
[in] tokenPresent = 0x1
[out] pSlotList: 
[out] *pulCount = 0x0
Returned:  0 CKR_OK

5: C_Finalize
2018-02-25 16:29:50.148
Returned:  0 CKR_OK
```

### Proposed Resolution

Get or write a token-specific driver. ;-)

I can't believe I'm the first one hitting this problem - is/was there an effort under way to get OpenSC working with the Latvian eID, which I assume is or at least should be similar to EU eID (I hope there is a EU standard? ;-)?

### Logs

See above.

**Comment by dengert** [comment]
@mouse07410  How did you get a Latvian eID card?

**Comment by metsma** [comment]
https://github.com/eID-LV/Middleware/tree/master/opensc-0.12.2

**Comment by mouse07410** [comment]
> How did you get a Latvian eID card?

Long story. One friend is a Latvian, and he recently got an eID there.

The real question is how to get it working, assuming it's possible. ;-)

**Comment by mouse07410** [comment]
> https://github.com/eID-LV/Middleware/tree/master/opensc-0.12.2

I **wish** the authors documented their stuff, so one could port their additions to the current OpenSC.

**Comment by dengert** [comment]
I have a opensc-0.12.2.tar.gz file if you want to see what they changed.

**Comment by dengert** [comment]
OpenSC git has a tag for version 0.12.0 so you could check it out. and do the diffs.

**Comment by dengert** [comment]
Sorry, a tag for 0.12.2

**Comment by mouse07410** [comment]
> OpenSC git has a tag for version 0.12.0 so you could check it out. and do the diffs

I'm not a Git ninja... :-(

So do I download the original OpenSC and checkout 0.12.0, then download their fork, and do directory compare or something like that?

**Comment by dengert** [comment]
The versions are 0.12.2. 

Start here:
https://stackoverflow.com/questions/35979642/how-to-checkout-remote-git-tag/35979751

In your git respository, run:
`git fetch --all --tags --prune`
This gets all that tags from all the repositories you can access. 
I my case `upstream` is the OpenSC/OpenSC  repos, and `origin` is my dengert/OpenSC repro.  

`git tag --list` 
shows all the tags available.
There is both a 0.12.2 and a v0.12.2 I think we want 0.12.2

`git checkout tags/0.12.2 -b tag-0.12.2`
`Switched to a new branch 'tag-0.12.2'`
which shows I have a new branch `tag-0.12.2` which matches both tags 0.12.2 and v0.12.2
on commit d505a55483f830bb612e1ff6c7fb1c65b7a9ad81  2011-07-15

We will make a new directory somewhere, call it `/tmp/eID-LV` for example. 
From https://github.com/eID-LV/Middleware
You can click on "clone or download" button, and copy `git@github.com:eID-LV/Middleware.git` then  
```
cd /tmp
mkdir eID-LV`
cd /tmpeID-LV
git clone git@github.com:eID-LV/Middleware.git 
cd Middleware
cd opensc-0.12.2
```
That puts you into their version of 0.12.2 in my case `/tmp/eID-LV/Middleware/opensc-0.12.2`

You can the compare directories recursively: `diff -u -r <dir1> <dir2>` 
`diff -u -r <your checked-out-version-of OpenSC-0.12.2>/src/libopensc  /tmp/eID-LV/Middleware/opensc-0.12.2/src/libopensc`

They may have change other things too.  Disregard much of the build changes  that may be different.

**Comment by frankmorgner** [comment]
If you're planning to make a contribution to the project, we'd love to add a new card to OpenSC. We would be happy to get you going and to comment on your code. However, please have in mind that in lack of hardware, documentation and manpower, we can't do this on our own.

Looking at the changes from https://github.com/eID-LV/Middleware/tree/master/opensc-0.12.2 would be a start, but basically you're on your own in terms of testing and implementing.

**Comment by mouse07410** [comment]
Thanks to @dengert, I've a better idea what they changed in 0.12.2:
```
$ diff -u -r OpenSC-eID/src eID-LV/opensc-0.12.2/src | fgrep diff
diff -u -r OpenSC-eID/src/common/simclist.c eID-LV/opensc-0.12.2/src/common/simclist.c
diff -u -r OpenSC-eID/src/libopensc/Makefile.am eID-LV/opensc-0.12.2/src/libopensc/Makefile.am
diff -u -r OpenSC-eID/src/libopensc/card-authentic.c eID-LV/opensc-0.12.2/src/libopensc/card-authentic.c
diff -u -r OpenSC-eID/src/libopensc/card-iasecc.c eID-LV/opensc-0.12.2/src/libopensc/card-iasecc.c
diff -u -r OpenSC-eID/src/libopensc/card-rtecp.c eID-LV/opensc-0.12.2/src/libopensc/card-rtecp.c
diff -u -r OpenSC-eID/src/libopensc/card-rutoken.c eID-LV/opensc-0.12.2/src/libopensc/card-rutoken.c
diff -u -r OpenSC-eID/src/libopensc/card.c eID-LV/opensc-0.12.2/src/libopensc/card.c
diff -u -r OpenSC-eID/src/libopensc/cards.h eID-LV/opensc-0.12.2/src/libopensc/cards.h
diff -u -r OpenSC-eID/src/libopensc/ctx.c eID-LV/opensc-0.12.2/src/libopensc/ctx.c
diff -u -r OpenSC-eID/src/libopensc/log.c eID-LV/opensc-0.12.2/src/libopensc/log.c
diff -u -r OpenSC-eID/src/libopensc/opensc.h eID-LV/opensc-0.12.2/src/libopensc/opensc.h
diff -u -r OpenSC-eID/src/libopensc/pkcs15-pin.c eID-LV/opensc-0.12.2/src/libopensc/pkcs15-pin.c
diff -u -r OpenSC-eID/src/libopensc/pkcs15.h eID-LV/opensc-0.12.2/src/libopensc/pkcs15.h
diff -u -r OpenSC-eID/src/libopensc/reader-pcsc.c eID-LV/opensc-0.12.2/src/libopensc/reader-pcsc.c
diff -u -r OpenSC-eID/src/libopensc/sec.c eID-LV/opensc-0.12.2/src/libopensc/sec.c
diff -u -r OpenSC-eID/src/minidriver/Makefile.am eID-LV/opensc-0.12.2/src/minidriver/Makefile.am
diff -u -r OpenSC-eID/src/pkcs11/Makefile.am eID-LV/opensc-0.12.2/src/pkcs11/Makefile.am
diff -u -r OpenSC-eID/src/pkcs11/framework-pkcs15.c eID-LV/opensc-0.12.2/src/pkcs11/framework-pkcs15.c
diff -u -r OpenSC-eID/src/pkcs11/mechanism.c eID-LV/opensc-0.12.2/src/pkcs11/mechanism.c
diff -u -r OpenSC-eID/src/pkcs11/misc.c eID-LV/opensc-0.12.2/src/pkcs11/misc.c
diff -u -r OpenSC-eID/src/pkcs11/pkcs11-object.c eID-LV/opensc-0.12.2/src/pkcs11/pkcs11-object.c
diff -u -r OpenSC-eID/src/pkcs11/sc-pkcs11.h eID-LV/opensc-0.12.2/src/pkcs11/sc-pkcs11.h
diff -u -r OpenSC-eID/src/pkcs11/slot.c eID-LV/opensc-0.12.2/src/pkcs11/slot.c
diff -u -r OpenSC-eID/src/pkcs15init/pkcs15-iasecc.c eID-LV/opensc-0.12.2/src/pkcs15init/pkcs15-iasecc.c
diff -u -r OpenSC-eID/src/pkcs15init/profile.c eID-LV/opensc-0.12.2/src/pkcs15init/profile.c
diff -u -r OpenSC-eID/src/tests/Makefile.am eID-LV/opensc-0.12.2/src/tests/Makefile.am
diff -u -r OpenSC-eID/src/tools/Makefile.am eID-LV/opensc-0.12.2/src/tools/Makefile.am
diff -u -r OpenSC-eID/src/tools/opensc-explorer.c eID-LV/opensc-0.12.2/src/tools/opensc-explorer.c
diff -u -r OpenSC-eID/src/tools/pkcs15-tool.c eID-LV/opensc-0.12.2/src/tools/pkcs15-tool.c
$ 
```

I'm seeing differences line this:
```
diff -u -r OpenSC-eID/src/libopensc/sec.c eID-LV/opensc-0.12.2/src/libopensc/sec.c
--- OpenSC-eID/src/libopensc/sec.c	2018-02-26 11:55:06.000000000 -0500
+++ eID-LV/opensc-0.12.2/src/libopensc/sec.c	2018-02-26 11:56:22.000000000 -0500
@@ -123,7 +123,7 @@
 	return sc_pin_cmd(card, &data, tries_left);
 }
 
-int sc_reset_retry_counter(sc_card_t *card, unsigned int type, int ref,
+int sc_reset_retry_counter(sc_card_t *card, unsigned int type, int so_ref, int ref,
 			   const u8 *puk, size_t puklen, const u8 *newref,
 			   size_t newlen)
 {
@@ -133,6 +133,7 @@
 	data.cmd = SC_PIN_CMD_UNBLOCK;
 	data.pin_type = type;
 	data.pin_reference = ref;
+    data.so_pin_reference = so_ref;
 	data.pin1.data = puk;
 	data.pin1.len = puklen;
 	data.pin2.data = newref;
@@ -187,6 +188,7 @@
 			if (card->ops->reset_retry_counter != NULL)
 				r = card->ops->reset_retry_counter(card,
 					data->pin_type,
+                    data->so_pin_reference,
 					data->pin_reference,
 					data->pin1.data,
 					(size_t) data->pin1.len,
```

The current master has `so_pin` and related stuff only in `card-sc-hsm.c` and `pkcs15.c` files. Are we willing to add this capability (broader support for `so_pin`) and the changes it brings in...?

Here's the complete list of diff's: 
[eID-LV.diff.txt](https://github.com/OpenSC/OpenSC/files/1759939/eID-LV.diff.txt)

**Comment by dengert** [comment]
There my be other changes like  changes to opensc.conf that are not added
to source. They had a lot of stuff in their git repro outside opensc.

**Comment by mouse07410** [comment]
> There my be other changes like  changes to opensc.conf that are not added
to source. They had a lot of stuff in their git repro outside opensc

Yes I realize that. My hope is that if `libopensc` recognizes the card and can work with it, then both `libp11` and `tokend` would be able to work with it - and that would cover all of my needs.

If I'm missing something - I'd appreciate if you could point me at it.

**Comment by frankmorgner** [comment]
OpenSC doesn't lack support for SO PINs. It's just not implemented the way @eID-LV hacked it into the code.

You can have multiple PIN objects in OpenSC. SO PIN is typically used to unblock a user PIN. So in the (emulated) PKCS#15 profile you'd have to add the user PIN's Auth ID to the ID of the SO PIN.

**Comment by frankmorgner** [comment]
Looking at the changes, I think you can't reuse any of the existing code. Basically @eID-LV added a number of ugly hacks into OpenSC to make their card work, while completely breaking existing support of other cards.

**Comment by frankmorgner** [comment]
If you need more input while implementing the changes, please reopen this issue

**Comment by muktupavels** [comment]
@frankmorgner Any suggestions how to properly add support for Latvian eID card?

**Comment by mouse07410** [comment]
In order to get Latvian eID card supported, at the very least one would need to know what kind of card it is, what kind of applet it runs, and what kind of keys/algorithms (RSA, ECC, or...?) it deals with.

So far I've been unable to get that info from the web site presumably run by the department of Latvian government that's presumably responsible for eID. If you can get and post that info, something might be done.

**Comment by frankmorgner** [comment]
You should start writing a card driver from scratch and look into https://github.com/eID-LV/Middleware. For example, you can run some tests with https://github.com/eID-LV/Middleware to get APDU traces. Basically, you need to understand how the card is used, there's no easy way around this.

**Comment by muktupavels** [comment]
For experimenting and testing I have this:
```
diff --git a/src/libopensc/card-iasecc.c b/src/libopensc/card-iasecc.c
index 8f5be901..c9f39906 100644
--- a/src/libopensc/card-iasecc.c
+++ b/src/libopensc/card-iasecc.c
@@ -74,8 +74,10 @@ static const struct sc_atr_table iasecc_known_atrs[] = {
 	{ "3B:7F:96:00:00:00:31:B8:64:40:70:14:10:73:94:01:80:82:90:00",
 	  "FF:FF:FF:FF:FF:FF:FF:FE:FF:FF:00:00:FF:FF:FF:FF:FF:FF:FF:FF",
 		"IAS/ECC Gemalto", SC_CARD_TYPE_IASECC_GEMALTO,  0, NULL },
-        { "3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:08:00:07:90:00:FE", NULL,
+	{ "3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:08:00:07:90:00:FE", NULL,
 		"IAS/ECC v1.0.1 Oberthur", SC_CARD_TYPE_IASECC_OBERTHUR,  0, NULL },
+	{ "3B:DD:18:00:81:31:FE:45:90:4C:41:54:56:49:41:2D:65:49:44:90:00:8C", NULL,
+		"IAS/ECC v1.0.1 Latvia eID", SC_CARD_TYPE_IASECC_OBERTHUR_LATEID, 0, NULL },
 	{ "3B:7D:13:00:00:4D:44:57:2D:49:41:53:2D:43:41:52:44:32", NULL,
 		"IAS/ECC v1.0.1 Sagem MDW-IAS-CARD2", SC_CARD_TYPE_IASECC_SAGEM,  0, NULL },
 	{ "3B:7F:18:00:00:00:31:B8:64:50:23:EC:C1:73:94:01:80:82:90:00", NULL,
@@ -507,6 +509,75 @@ iasecc_init_oberthur(struct sc_card *card)
 }
 
 
+static int
+iasecc_oberthur_match_lateid(struct sc_card *card)
+{
+	struct sc_context *ctx = card->ctx;
+	//unsigned char *hist = card->reader->atr_info.hist_bytes;
+
+	LOG_FUNC_CALLED(ctx);
+
+#if 0
+	if (*hist != 0x80 || ((*(hist+1)&0xF0) != 0xF0))
+		LOG_FUNC_RETURN(ctx, SC_ERROR_OBJECT_NOT_FOUND);
+
+	sc_log_hex(ctx, "AID in historical_bytes", hist + 2, *(hist+1) & 0x0F);
+
+	if (memcmp(hist + 2, OberthurIASECC_AID.value, *(hist+1) & 0x0F))
+		LOG_FUNC_RETURN(ctx, SC_ERROR_RECORD_NOT_FOUND);
+#endif
+
+	if (!card->ef_atr)
+		card->ef_atr = calloc(1, sizeof(struct sc_ef_atr));
+	if (!card->ef_atr)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_OUT_OF_MEMORY);
+
+	memcpy(card->ef_atr->aid.value, OberthurIASECC_AID.value, OberthurIASECC_AID.len);
+	card->ef_atr->aid.len = OberthurIASECC_AID.len;
+
+	LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+}
+
+
+static int
+iasecc_init_oberthur_lateid(struct sc_card *card)
+{
+	struct sc_context *ctx = card->ctx;
+	unsigned int flags;
+	int rv = 0;
+
+	LOG_FUNC_CALLED(ctx);
+
+	flags = IASECC_CARD_DEFAULT_FLAGS;
+
+	_sc_card_add_rsa_alg(card, 1024, flags, 0x10001);
+	_sc_card_add_rsa_alg(card, 2048, flags, 0x10001);
+
+	card->caps = SC_CARD_CAP_RNG;
+	card->caps |= SC_CARD_CAP_APDU_EXT;
+	card->caps |= SC_CARD_CAP_USE_FCI_AC;
+
+	iasecc_parse_ef_atr(card);
+
+	/* if we fail to select CM, */
+	if (gp_select_card_manager(card)) {
+		gp_select_isd_rid(card);
+	}
+
+	rv = iasecc_oberthur_match_lateid(card);
+	LOG_TEST_RET(ctx, rv, "unknown Oberthur's IAS/ECC card");
+
+	rv = iasecc_select_mf(card, NULL);
+	LOG_TEST_RET(ctx, rv, "MF selection error");
+
+	rv = iasecc_parse_ef_atr(card);
+	LOG_TEST_RET(ctx, rv, "EF.ATR read or parse error");
+
+	sc_log(ctx, "EF.ATR(aid:'%s')", sc_dump_hex(card->ef_atr->aid.value, card->ef_atr->aid.len));
+	LOG_FUNC_RETURN(ctx, rv);
+}
+
+
 static int
 iasecc_mi_match(struct sc_card *card)
 {
@@ -591,6 +662,8 @@ iasecc_init(struct sc_card *card)
 		rv = iasecc_init_gemalto(card);
 	else if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR)
 		rv = iasecc_init_oberthur(card);
+	else if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR_LATEID)
+		rv = iasecc_init_oberthur_lateid(card);
 	else if (card->type == SC_CARD_TYPE_IASECC_SAGEM)
 		rv = iasecc_init_amos_or_sagem(card);
 	else if (card->type == SC_CARD_TYPE_IASECC_AMOS)
@@ -898,6 +971,7 @@ iasecc_select_file(struct sc_card *card, const struct sc_path *path,
 
 		if (card->type != SC_CARD_TYPE_IASECC_GEMALTO
 				&& card->type != SC_CARD_TYPE_IASECC_OBERTHUR
+				&& card->type != SC_CARD_TYPE_IASECC_OBERTHUR_LATEID
 				&& card->type != SC_CARD_TYPE_IASECC_SAGEM
 				&& card->type != SC_CARD_TYPE_IASECC_AMOS
 				&& card->type != SC_CARD_TYPE_IASECC_MI
@@ -910,6 +984,10 @@ iasecc_select_file(struct sc_card *card, const struct sc_path *path,
 				apdu.p1 = 0x01;
 				apdu.p2 = 0x04;
 			}
+			if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR_LATEID)   {
+				apdu.p1 = 0x01;
+				apdu.p2 = 0x04;
+			}
 			if (card->type == SC_CARD_TYPE_IASECC_AMOS)
 				apdu.p2 = 0x04;
 			if (card->type == SC_CARD_TYPE_IASECC_MI)
@@ -921,6 +999,8 @@ iasecc_select_file(struct sc_card *card, const struct sc_path *path,
 			apdu.p1 = 0x09;
 			if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR)
 				apdu.p2 = 0x04;
+			if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR_LATEID)
+				apdu.p2 = 0x04;
 			if (card->type == SC_CARD_TYPE_IASECC_AMOS)
 				apdu.p2 = 0x04;
 			if (card->type == SC_CARD_TYPE_IASECC_MI)
diff --git a/src/libopensc/cards.h b/src/libopensc/cards.h
index 121182bb..84188d3f 100644
--- a/src/libopensc/cards.h
+++ b/src/libopensc/cards.h
@@ -204,6 +204,7 @@ enum {
 	SC_CARD_TYPE_IASECC_BASE = 25000,
 	SC_CARD_TYPE_IASECC_GEMALTO,
 	SC_CARD_TYPE_IASECC_OBERTHUR,
+	SC_CARD_TYPE_IASECC_OBERTHUR_LATEID,
 	SC_CARD_TYPE_IASECC_SAGEM,
 	SC_CARD_TYPE_IASECC_AMOS,
 	SC_CARD_TYPE_IASECC_MI,
```

It is enough to get card recognized and some things seems to work (tested with my own written code) - GetInfo, GetSlotList, GetSlotInfo, GetTokenInfo, OpenSession, GetSessionInfo, FindObjectsInit, FindObjects, GetAttributeValue, FindObjectsFinal, CloseSession. One more tested thing is Login function that seems to not work. It says that I have entered wrong pin. After that remaining tries are properly decreased from 3 to 2.

Is it possible to get at least some useful info from card? What command(s) should I run.

What is APDU traces? And how can I get them?

**Comment by muktupavels** [comment]
With `systemctl stop pcscd` followed by `sudo LIBCCID_ifdLogLevel=0x000F pcscd --foreground --debug` I was able to see sent data. Here is example for C_Login function:

> **otlv-pkcs11.so:**
> 00000017 [140689321740032] commands.c:2228:CmdXfrBlockTPDU_T1() T=1: 69 and 65548 bytes
> 00000015 [140689321740032] openct/proto-t1.c:580:t1_build() more bit: 0
> 00000035 [140689321740032] sending: 00 00 45 00 20 00 01 40 XX XX XX XX FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 27 
> 00000041 [140689321740032] -> 000000 6F 49 00 00 00 00 A3 00 00 00 00 00 45 00 20 00 01 40 XX XX XX XX FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF 27 
> 00043212 [140689321740032] <- 000000 80 06 00 00 00 00 A3 00 00 00 00 00 02 90 00 92 
> 00000034 [140689321740032] received: 00 00 02 90 00 92 
> 00000018 [140689321740032] openct/proto-t1.c:356:t1_transceive()
> 
> **opensc-pkcs11.so:**
> 00000012 [140689321740032] commands.c:2228:CmdXfrBlockTPDU_T1() T=1: 9 and 65548 bytes
> 00000013 [140689321740032] openct/proto-t1.c:580:t1_build() more bit: 0
> 00000011 [140689321740032] sending: 00 00 09 00 20 00 01 04 XX XX XX XX 2F 
> 00000018 [140689321740032] -> 000000 6F 0D 00 00 00 00 37 00 00 00 00 00 09 00 20 00 01 04 XX XX XX XX 2F 
> 00029938 [140689321740032] <- 000000 80 06 00 00 00 00 37 00 00 00 00 00 02 63 C2 A3 
> 00000018 [140689321740032] received: 00 00 02 63 C2 A3 
> 00000005 [140689321740032] openct/proto-t1.c:356:t1_transceive() 

```
diff --git a/src/libopensc/card-iasecc.c b/src/libopensc/card-iasecc.c
index 8f5be901..1aa7df4a 100644
--- a/src/libopensc/card-iasecc.c
+++ b/src/libopensc/card-iasecc.c
@@ -74,8 +74,10 @@ static const struct sc_atr_table iasecc_known_atrs[] = {
 	{ "3B:7F:96:00:00:00:31:B8:64:40:70:14:10:73:94:01:80:82:90:00",
 	  "FF:FF:FF:FF:FF:FF:FF:FE:FF:FF:00:00:FF:FF:FF:FF:FF:FF:FF:FF",
 		"IAS/ECC Gemalto", SC_CARD_TYPE_IASECC_GEMALTO,  0, NULL },
-        { "3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:08:00:07:90:00:FE", NULL,
+	{ "3B:DD:18:00:81:31:FE:45:80:F9:A0:00:00:00:77:01:08:00:07:90:00:FE", NULL,
 		"IAS/ECC v1.0.1 Oberthur", SC_CARD_TYPE_IASECC_OBERTHUR,  0, NULL },
+	{ "3B:DD:18:00:81:31:FE:45:90:4C:41:54:56:49:41:2D:65:49:44:90:00:8C", NULL,
+		"IAS/ECC v1.0.1 Latvia eID", SC_CARD_TYPE_IASECC_OBERTHUR_LATEID, 0, NULL },
 	{ "3B:7D:13:00:00:4D:44:57:2D:49:41:53:2D:43:41:52:44:32", NULL,
 		"IAS/ECC v1.0.1 Sagem MDW-IAS-CARD2", SC_CARD_TYPE_IASECC_SAGEM,  0, NULL },
 	{ "3B:7F:18:00:00:00:31:B8:64:50:23:EC:C1:73:94:01:80:82:90:00", NULL,
@@ -507,6 +509,75 @@ iasecc_init_oberthur(struct sc_card *card)
 }
 
 
+static int
+iasecc_oberthur_match_lateid(struct sc_card *card)
+{
+	struct sc_context *ctx = card->ctx;
+	//unsigned char *hist = card->reader->atr_info.hist_bytes;
+
+	LOG_FUNC_CALLED(ctx);
+
+#if 0
+	if (*hist != 0x80 || ((*(hist+1)&0xF0) != 0xF0))
+		LOG_FUNC_RETURN(ctx, SC_ERROR_OBJECT_NOT_FOUND);
+
+	sc_log_hex(ctx, "AID in historical_bytes", hist + 2, *(hist+1) & 0x0F);
+
+	if (memcmp(hist + 2, OberthurIASECC_AID.value, *(hist+1) & 0x0F))
+		LOG_FUNC_RETURN(ctx, SC_ERROR_RECORD_NOT_FOUND);
+#endif
+
+	if (!card->ef_atr)
+		card->ef_atr = calloc(1, sizeof(struct sc_ef_atr));
+	if (!card->ef_atr)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_OUT_OF_MEMORY);
+
+	memcpy(card->ef_atr->aid.value, OberthurIASECC_AID.value, OberthurIASECC_AID.len);
+	card->ef_atr->aid.len = OberthurIASECC_AID.len;
+
+	LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+}
+
+
+static int
+iasecc_init_oberthur_lateid(struct sc_card *card)
+{
+	struct sc_context *ctx = card->ctx;
+	unsigned int flags;
+	int rv = 0;
+
+	LOG_FUNC_CALLED(ctx);
+
+	flags = IASECC_CARD_DEFAULT_FLAGS;
+
+	_sc_card_add_rsa_alg(card, 1024, flags, 0x10001);
+	_sc_card_add_rsa_alg(card, 2048, flags, 0x10001);
+
+	card->caps = SC_CARD_CAP_RNG;
+	card->caps |= SC_CARD_CAP_APDU_EXT;
+	card->caps |= SC_CARD_CAP_USE_FCI_AC;
+
+	iasecc_parse_ef_atr(card);
+
+	/* if we fail to select CM, */
+	if (gp_select_card_manager(card)) {
+		gp_select_isd_rid(card);
+	}
+
+	rv = iasecc_oberthur_match_lateid(card);
+	LOG_TEST_RET(ctx, rv, "unknown Oberthur's IAS/ECC card");
+
+	rv = iasecc_select_mf(card, NULL);
+	LOG_TEST_RET(ctx, rv, "MF selection error");
+
+	rv = iasecc_parse_ef_atr(card);
+	LOG_TEST_RET(ctx, rv, "EF.ATR read or parse error");
+
+	sc_log(ctx, "EF.ATR(aid:'%s')", sc_dump_hex(card->ef_atr->aid.value, card->ef_atr->aid.len));
+	LOG_FUNC_RETURN(ctx, rv);
+}
+
+
 static int
 iasecc_mi_match(struct sc_card *card)
 {
@@ -591,6 +662,8 @@ iasecc_init(struct sc_card *card)
 		rv = iasecc_init_gemalto(card);
 	else if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR)
 		rv = iasecc_init_oberthur(card);
+	else if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR_LATEID)
+		rv = iasecc_init_oberthur_lateid(card);
 	else if (card->type == SC_CARD_TYPE_IASECC_SAGEM)
 		rv = iasecc_init_amos_or_sagem(card);
 	else if (card->type == SC_CARD_TYPE_IASECC_AMOS)
@@ -898,6 +971,7 @@ iasecc_select_file(struct sc_card *card, const struct sc_path *path,
 
 		if (card->type != SC_CARD_TYPE_IASECC_GEMALTO
 				&& card->type != SC_CARD_TYPE_IASECC_OBERTHUR
+				&& card->type != SC_CARD_TYPE_IASECC_OBERTHUR_LATEID
 				&& card->type != SC_CARD_TYPE_IASECC_SAGEM
 				&& card->type != SC_CARD_TYPE_IASECC_AMOS
 				&& card->type != SC_CARD_TYPE_IASECC_MI
@@ -910,6 +984,10 @@ iasecc_select_file(struct sc_card *card, const struct sc_path *path,
 				apdu.p1 = 0x01;
 				apdu.p2 = 0x04;
 			}
+			if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR_LATEID)   {
+				apdu.p1 = 0x01;
+				apdu.p2 = 0x04;
+			}
 			if (card->type == SC_CARD_TYPE_IASECC_AMOS)
 				apdu.p2 = 0x04;
 			if (card->type == SC_CARD_TYPE_IASECC_MI)
@@ -921,6 +999,8 @@ iasecc_select_file(struct sc_card *card, const struct sc_path *path,
 			apdu.p1 = 0x09;
 			if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR)
 				apdu.p2 = 0x04;
+			if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR_LATEID)
+				apdu.p2 = 0x04;
 			if (card->type == SC_CARD_TYPE_IASECC_AMOS)
 				apdu.p2 = 0x04;
 			if (card->type == SC_CARD_TYPE_IASECC_MI)
@@ -1859,9 +1939,22 @@ iasecc_chv_verify(struct sc_card *card, struct sc_pin_cmd_data *pin_cmd,
 	}
 	else if (pin_cmd->pin1.data && pin_cmd->pin1.len)   {
 		sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x20, 0, pin_cmd->pin_reference);
-		apdu.data = pin_cmd->pin1.data;
-		apdu.datalen = pin_cmd->pin1.len;
-		apdu.lc = pin_cmd->pin1.len;
+
+		if (pin_cmd->pin1.pad_length && pin_cmd->flags & SC_PIN_CMD_NEED_PADDING)   {
+			unsigned char pin_buff[SC_MAX_APDU_BUFFER_SIZE] = {0};
+
+			memcpy(pin_buff, pin_cmd->pin1.data, pin_cmd->pin1.len);
+			memset(pin_buff + pin_cmd->pin1.len, pin_cmd->pin1.pad_char, pin_cmd->pin1.pad_length - pin_cmd->pin1.len);
+
+			apdu.data = pin_buff;
+			apdu.datalen = pin_cmd->pin1.pad_length;
+			apdu.lc = pin_cmd->pin1.pad_length;
+		}
+		else {
+			apdu.data = pin_cmd->pin1.data;
+			apdu.datalen = pin_cmd->pin1.len;
+			apdu.lc = pin_cmd->pin1.len;
+		}
 	}
 	else if ((card->reader->capabilities & SC_READER_CAP_PIN_PAD) && !pin_cmd->pin1.data && !pin_cmd->pin1.len)   {
 		rv = iasecc_chv_verify_pinpad(card, pin_cmd, tries_left);
@@ -2248,6 +2341,13 @@ iasecc_pin_get_policy (struct sc_card *card, struct sc_pin_cmd_data *data)
 	data->pin1.offset = 5;
 	data->pin1.logged_in = SC_PIN_STATE_UNKNOWN;
 
+	if (card->type == SC_CARD_TYPE_IASECC_OBERTHUR_LATEID) {
+		data->pin1.pad_char = 0xFF;
+		data->pin1.pad_length = data->pin1.max_length;
+
+		data->flags |= SC_PIN_CMD_NEED_PADDING;
+	}
+
 	sc_log(ctx,
 	       "PIN policy: size max/min %"SC_FORMAT_LEN_SIZE_T"u/%"SC_FORMAT_LEN_SIZE_T"u, tries max/left %i/%i",
 	       data->pin1.max_length, data->pin1.min_length,
diff --git a/src/libopensc/cards.h b/src/libopensc/cards.h
index 121182bb..84188d3f 100644
--- a/src/libopensc/cards.h
+++ b/src/libopensc/cards.h
@@ -204,6 +204,7 @@ enum {
 	SC_CARD_TYPE_IASECC_BASE = 25000,
 	SC_CARD_TYPE_IASECC_GEMALTO,
 	SC_CARD_TYPE_IASECC_OBERTHUR,
+	SC_CARD_TYPE_IASECC_OBERTHUR_LATEID,
 	SC_CARD_TYPE_IASECC_SAGEM,
 	SC_CARD_TYPE_IASECC_AMOS,
 	SC_CARD_TYPE_IASECC_MI,
```

With this `C_Login` accepts my PIN and PUK codes. But I can not change / set new pin code. I guess it is same problem!? New PIN codes should also be padded with 0xff.

Right now it seems that Latvian eID card is very similar to `SC_CARD_TYPE_IASECC_OBERTHUR` with following differences:
- Different historical bytes
- PIN codes needs to be padded with 0xff

Would above diff be right direction to add support for it? I guess there is no real need for new `*_lateid` functions that mostly duplicates code. `iasecc_oberthur_match` probably can be updated to also accept Latvian eID cards historical bytes.
